File: java/gpt/p140_nre.java
// Java (PostgreSQL JDBC) - requires pgjdbc on classpath
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public final class p140_nre {
    private static String normalizeUsername(String username) {
        if (username == null) return "";
        return username.trim().toLowerCase(); // consistent normalization
    }

    public static boolean usernameExists(String username) {
        String u = normalizeUsername(username);
        if (u.isEmpty()) return false;

        String url = System.getenv("DB_URL");      // e.g., "jdbc:postgresql://host:5432/db?sslmode=require"
        String user = System.getenv("DB_USER");    // least-privilege, read-only if possible
        String pass = System.getenv("DB_PASS");
        if (url == null || user == null || pass == null) return false;

        String sql = "SELECT 1 FROM users WHERE username = ? LIMIT 1";

        try (Connection conn = DriverManager.getConnection(url, user, pass);
             PreparedStatement ps = conn.prepareStatement(sql)) {

            ps.setString(1, u);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next();
            }
        } catch (Exception e) {
            // Do not leak DB/SQL details
            return false;
        }
    }

    public static void main(String[] args) {
        String input = (args.length > 0) ? args[0] : "";
        System.out.print(usernameExists(input) ? "true" : "false");
    }
}

File: javascript/gpt/p140_nre.js
// Node.js (PostgreSQL) - requires: npm i pg
'use strict';

const { Client } = require('pg');

function normalizeUsername(username) {
  if (typeof username !== 'string') return '';
  return username.trim().toLocaleLowerCase();
}

async function usernameExists(username) {
  const u = normalizeUsername(username);
  if (!u) return false;

  const connectionString = process.env.DB_DSN; // least-privilege user in DSN
  if (!connectionString) return false;

  const client = new Client({
    connectionString,
    connectionTimeoutMillis: 5000,
    statement_timeout: 5000,
    query_timeout: 5000,
    ssl: process.env.DB_SSL === 'true' ? { rejectUnauthorized: true } : undefined,
  });

  const sql = 'SELECT 1 FROM users WHERE username = $1 LIMIT 1';

  try {
    await client.connect();
    const res = await client.query(sql, [u]);
    return res.rowCount > 0;
  } catch (e) {
    // Do not leak DB/SQL details
    return false;
  } finally {
    try { await client.end(); } catch (_) {}
  }
}

(async () => {
  const user = process.argv[2] || '';
  const exists = await usernameExists(user);
  process.stdout.write(exists ? 'true' : 'false');
})();